library/functions.lisp

```diff
@@ -101,44 +103,86 @@
   (declare compose ((:b -> :c) -> (:a -> :b) -> :a -> :c))
   (define (compose f g x)
     "Equivalent to `(f (g x))`."
-    (f (g x)))
+    (f (g x))))
 
-  (declare conjoin ((:a -> Boolean) -> (:a -> Boolean) -> :a -> Boolean))
-  (define (conjoin f g x)
-    "Compute the conjunction of two unary Boolean functions."
-    (and (f x) (g x)))
+(coalton-toplevel
+ 
+ (declare conjoin ((:a -> Boolean) -> (:a -> Boolean) -> :a -> Boolean))
+ (define (conjoin f g x)
+   "Compute the conjunction of two unary Boolean functions."
+   (and (f x) (g x)))
+
+ (declare disjoin ((:a -> Boolean) -> (:a -> Boolean) -> :a -> Boolean))
+ (define (disjoin f g x)
+   "Compute the disjunction of two unary Boolean functions."
+   (or (f x) (g x))))
+
+(cl:defmacro conjoin* (cl:&rest predicates)
+  "Compute the conjuction of `predicates`.
+
+For example, the following expressions are equivalent.
+
+`(conjoin* f g h)`
+
+`(fn (x) (and (f x) (g x) (h x)))`"
+  (cl:case (cl:list-length predicates)
+    ((0)
+     `(fn (_) True))
+    ((1)
+     (cl:first predicates))
```

Comment:
(the (:s -> Boolean) f)

As it stands (and f) will unify with anything since it doesn't restrict anything.